controller: wfa_ca: implement initial testbed commands handlers

tomereli opened the pull request:
This PR adds implementation to DEV_RESET_DEFAULT, DEV_GET_INFO, DEV_GET_PARAMETER and adds a topology query handler. All are needed by the first MCUT test case as defined in EasyMesh test plan 1.0 section 5.3.1.

Commit:
controller: wfa_ca: fix DEV_RESET_DEFAULT

The command sent to the DUT in DEV_RESET_DEFAULT is as follows (tested
with the formal UCC application):
INFO     DUT (127.0.0.1:5000) --->
dev_reset_default,name,BroadcomCTTDUT,program,map,devrole,agent,type,DUT

Fix wfa_ca.cpp parser to accept DUT instead of dut for type param.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: wfa_ca: Add DEV_GET_INFO implementaion

Add DEV_GET_INFO implementation which always returns:
STATUS,COMPLETE,vendor,Intel,model,prplMesh,version,1.4.0

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: wfa_ca: Add DEV_GET_PARAMETER impl

Add DEV_GET_PARAMETER implementation which on controller certification
receives the following string from the WTS application:
DUT (127.0.0.1:5000) ---> dev_get_parameter,program,map,parameter,ALid

And replies with the "br-lan" MAC address as the ALid:
DUT (127.0.0.1:5000) <-- status,COMPLETE,aLid,0a:b1:a5:36:ce:88

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: wfa_ca: add topology query handler

Add topology query CMDU creation for topology query message type in
wfa_ca.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -313,7 +313,7 @@ void wfa_ca::handle_wfa_ca_message(
 
         if (params.find("type") != params.end()) {
-            if (params["type"] != "test bed" && params["type"] != "dut") {
+            if (params["type"] != "test bed" && params["type"] != "DUT") {
```
Basically, the CAPI commands should be **not** case sensitive with some exceptions like SSID, so maybe it was better to downcase the content of params["type"]. But on the other hand, it will be ugly and if the UCC sends it all the time up-case, then I guess it is ok.

tomereli replied:
It sends it in uppercase

arnout replied:
OK for this one, but if we're going to add more of this kind of comparisons, I think we should implement a helper function `get_param(std::string type, std::string default = "")` that handles the condition and converts to lowercase.

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -240,7 +240,43 @@ void wfa_ca::handle_wfa_ca_message(
+        }
+
+        // Send back second reply
+        reply(sd, cmdu_tx, eWfaCaStatus::COMPLETE, "vendor,Intel,model,prplMesh,version,1.4.0");
```
I would replace the hard codded version to the macro BEEROCKS_VERSION.
see the implementation of CA_GET_VERSION.

arnout replied:
Not entirely related, but: we can't keep vendor as Intel. We will have to make it configurable (also in the WSC attributes etc.), either with a define or from the config file.

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -280,7 +282,42 @@ void wfa_ca::handle_wfa_ca_message(
+                                                          "macaddr", it implies that device should return backhaul STA's MAC address. SSID is also
+                                                          not required for the parameter "ALid"
+        * 
+        * Return Values: 
```
Missing param name "parameter" on the table.
As well the description exceeds 100 notes. Please fix this on other tables you added. I look better when open the code in github.

tomereli replied:
removed all comments since they violate license

morantr commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -280,7 +282,42 @@ void wfa_ca::handle_wfa_ca_message(
+    
+        std::string alid;
+        if (!net::network_utils::linux_iface_get_mac("br-lan", alid))
+            reply(sd, cmdu_tx, eWfaCaStatus::INVALID, "FAIL");
```
Since this failure is not related to command validity and occurs after you send status RUNNING, I should be ERROR instead of INVALID.
Also please add debug log, and break.

Commit:
docker: add nc support for runner containers

Netcat (nc) is a networking utility for reading from and writing to
network connections using TCP or UDP, and can be used to send commands
to the WFA-CA listener to simulate UCC application commands.
It is therefore useful to have it in the runner containers - so add it
to the Dockerfile.

While we're at it - add port number argument to run.sh (default 5000) so
we can also use netcat from the host when using host network instead of
prplmesh network.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on controller/src/beerocks/master/wfa_ca.cpp:
```diff
@@ -280,7 +282,42 @@ void wfa_ca::handle_wfa_ca_message(
+        }
+    
+        std::string alid;
+        if (!net::network_utils::linux_iface_get_mac("br-lan", alid))
```
We *really* need a better way to get the AL-MAC address... But that's not for this PR.

Commit:
controller: config: add vendor and model

Preparatory commit.
Add vendor and model to the controller configuration file which will be
used by WFA CA listener to return the vendor and model in
device_get_info command.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout reviewed (approved):
Typo in last commit message: topolohy -> topology

Moran made all the relevant change requests, so I'll approve already.

arnout commented:
I've made the requested changes and rebased. @morantr please re-review.
